### PR TITLE
node: Fix syncing of local-gateway-mode nodeport services

### PR DESCRIPTION
**Pull request by danwinship**
Not sure if this is the best fix... if we got rid of all the `ValidatePort` stuff in `getGatewayIPTRules` (which is unnecessary anyway since kube has already validated all of that) then `getGatewayIPTRules` becomes short enough that maybe it would make more sense to just inline it rather than having it run with two separate modes...

cc @alexanderConstantinescu @aojea @dcbw

### CODE REVIEW

**Comment by dcbw** [comment]
`/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/gateway_iptables.go:127:26: undeclared name: gatewayIP`

File: go-controller/pkg/node/gateway_iptables.go [hunk 1]
```diff
@@ -122,7 +122,7 @@ func getLocalGatewayInitRules(chain string, proto iptables.Protocol) []iptRule {
 	}
 }
 
-func getNodePortIPTRules(svcPort kapi.ServicePort, nodeIP *net.IPNet, gatewayIP string, targetPort int32) []iptRule {
+func getNodePortIPTRules(svcPort kapi.ServicePort, nodeIP *net.IPNet, targetIP string, targetPort int32) []iptRule {
 	var protocol iptables.Protocol
 	if utilnet.IsIPv6String(gatewayIP) {
```
Comment (dcbw) [inline go-controller/pkg/node/gateway_iptables.go:127]: gatewayIP -> targetIP?

File: go-controller/pkg/node/gateway_iptables.go [hunk 2]
```diff
@@ -326,7 +326,12 @@ func recreateIPTRules(table, chain string, keepIPTRules []iptRule) {
 	}
 }
 
-func getGatewayIPTRules(service *kapi.Service, nodePortTargetIP string, nodeIP *net.IPNet) []iptRule {
+// getGatewayIPTRules returns NodePort and ExternalIP iptables rules for service. If nodeIP is non-nil, then
+// only incoming traffic on that IP will be accepted for NodePort rules; otherwise incoming traffic on the NodePort
+// on all IPs will be accepted. If internalLoadBalancerIP is "", then NodePort traffic will be redirected to
+// the service port on the service's ClusterIP. Otherwise, it will be redirected to the NodePort on the
+// internalLoadBalancerIP.
+func getGatewayIPTRules(service *kapi.Service, internalLoadBalancerIP string, nodeIP *net.IPNet) []iptRule {
```
Comment (trozet) [inline go-controller/pkg/node/gateway_iptables.go:334]: I wouldn't call this internalLoadBalancerIP; that's a little ambiguous. It's really an ovn gateway router IP.

File: go-controller/pkg/node/gateway_iptables.go [hunk 3]
```diff
@@ -326,7 +326,12 @@ func recreateIPTRules(table, chain string, keepIPTRules []iptRule) {
 	}
 }
 
-func getGatewayIPTRules(service *kapi.Service, nodePortTargetIP string, nodeIP *net.IPNet) []iptRule {
+// getGatewayIPTRules returns NodePort and ExternalIP iptables rules for service. If nodeIP is non-nil, then
+// only incoming traffic on that IP will be accepted for NodePort rules; otherwise incoming traffic on the NodePort
+// on all IPs will be accepted. If internalLoadBalancerIP is "", then NodePort traffic will be redirected to
```
Comment (trozet) [inline go-controller/pkg/node/gateway_iptables.go:331]: s/redirected/DNAT'ed

File: go-controller/pkg/node/gateway_iptables.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_iptables.go:334]: @girishmg @danwinship why does shared gateway mode behave like this? Is it because of the br-local thing in shared gw mode? For local gw (assuming a node ip of 1.1.1.1, GR of 169.254.0.2) we have packet destined for  1.1.1.1 node port. DNAT in host from 1.1.1.1-> 169.254.0.2. Packet arrives at GR, then DNAT's directly to endpoint (never goes through cluster IP).

Can someone describe the shared behavior?

File: go-controller/pkg/node/gateway_shared_intf_linux.go [hunk 4]
```diff
@@ -81,14 +81,14 @@ func setupLocalNodeAccessBridge(nodeName string, subnets []*net.IPNet) error {
 }
 
 func addSharedGatewayIptRules(service *kapi.Service, nodeIP *net.IPNet) {
-	rules := getGatewayIPTRules(service, service.Spec.ClusterIP, nodeIP)
+	rules := getGatewayIPTRules(service, "", nodeIP)
```
Comment (alexanderConstantinescu) [inline go-controller/pkg/node/gateway_shared_intf_linux.go:84]: Why is this necessary though? The condition in `getGatewayIPTRules` will force the usage of `service.Spec.ClusterIP` anyways.

File: go-controller/pkg/node/gateway_iptables.go (hunk 2 above)
Comment (alexanderConstantinescu) [reply go-controller/pkg/node/gateway_iptables.go:334]: Here is a bit of context from a discussion I had with @girishmg : https://ovn-org.slack.com/archives/C010SQ5FSNL/p1591295321264600

I don't think it answers all questions, but might help fill in some blanks

**Comment by alexanderConstantinescu** [comment]
I don't fully understand what this patch fixes. The title mentions: "syncing of local-gateway-mode nodeport services", but https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway_localnet.go#L349 already validates that `gatewayIP` has a defined value. I don't see any potential issues with it that would be solved by this patch.

File: go-controller/pkg/node/gateway_iptables.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_iptables.go:334]: Thanks. After talking with @girishmg in shared gw mode we DNAT from host directly to cluster IP, then route cluster IP to the distributed router and SNAT via MP0 (to guarantee return traffic path). This is because the GR has the same IP as the node, so this makes sense...for now. In the future we need to look at all of these paths into the data plane and see if we can consolidate them.

**Comment by danwinship** [comment]
@alexanderConstantinescu The bug is that `getGatewayIPTRules` currently always redirects incoming nodeport traffic to `svcPort.Port` on `nodePortTargetIP`, but for local gateway mode it needs to redirect it to `svcPort.NodePort`, not `svcPort.Port`. Since the port/nodeport vary per service port, the caller can't just pass in the port to use like we do with the IP, so I changed the semantics of `getGatewayIPTRules` so that if you pass in an IP, it will redirect to `svcPort.NodePort` on that IP, and if you _don't_ pass in an IP, it will redirect to `svcPort.Port` on `svc.Spec.ClusterIP`.

**Comment by dcbw** [comment]
(unrelated?) testcase race failure

```
2020-07-26T17:21:31.7829806Z ==================
2020-07-26T17:21:31.7830113Z WARNING: DATA RACE
2020-07-26T17:21:31.7830448Z Write at 0x0000032b6500 by goroutine 16:
2020-07-26T17:21:31.7831081Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config.PrepareTestConfig()
2020-07-26T17:21:31.7831910Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:343 +0x1a0
2020-07-26T17:21:31.7832532Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.glob..func5.1()
2020-07-26T17:21:31.7833207Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/master_test.go:247 +0x4f
2020-07-26T17:21:31.7833609Z   github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
2020-07-26T17:21:31.7834387Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113 +0xdf
2020-07-26T17:21:31.7834795Z   github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
2020-07-26T17:21:31.7835547Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:64 +0x107
2020-07-26T17:21:31.7835961Z   github.com/onsi/ginkgo/internal/leafnodes.(*SetupNode).Run()
2020-07-26T17:21:31.7836707Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/setup_nodes.go:15 +0x87
2020-07-26T17:21:31.7837066Z   github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
2020-07-26T17:21:31.7837782Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/spec/spec.go:193 +0x2c5
2020-07-26T17:21:31.7838177Z   github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
2020-07-26T17:21:31.7838894Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/spec/spec.go:138 +0x167
2020-07-26T17:21:31.7866054Z   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
2020-07-26T17:21:31.7920136Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:200 +0x172
2020-07-26T17:21:31.7920411Z   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
2020-07-26T17:21:31.7920998Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:170 +0x216
2020-07-26T17:21:31.7921262Z   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
2020-07-26T17:21:31.7921839Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:66 +0x140
2020-07-26T17:21:31.7922057Z   github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
2020-07-26T17:21:31.7922616Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/suite/suite.go:62 +0x69f
2020-07-26T17:21:31.7922845Z   github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
2020-07-26T17:21:31.7923510Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:226 +0x338
2020-07-26T17:21:31.7923759Z   github.com/onsi/ginkgo.RunSpecsWithDefaultAndCustomReporters()
2020-07-26T17:21:31.7924297Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:214 +0x11f
2020-07-26T17:21:31.7924503Z   github.com/onsi/ginkgo.RunSpecs()
2020-07-26T17:21:31.7925019Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:205 +0x1a2
2020-07-26T17:21:31.7925446Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.TestClusterNode()
2020-07-26T17:21:31.7926088Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/ovn_suite_test.go:12 +0xf8
2020-07-26T17:21:31.7926263Z   testing.tRunner()
2020-07-26T17:21:31.7926477Z       /opt/hostedtoolcache/go/1.13.4/x64/src/testing/testing.go:909 +0x199
2020-07-26T17:21:31.7926577Z 
2020-07-26T17:21:31.7926766Z Previous read at 0x0000032b6500 by goroutine 93:
2020-07-26T17:21:31.7927240Z   github.com/ovn-org/ovn-kubernetes/go-controller/hybrid-overlay/pkg/util.IsHybridOverlayNode()
2020-07-26T17:21:31.7927749Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/util/util.go:37 +0x3e
2020-07-26T17:21:31.7928339Z   github.com/ovn-org/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller.(*MasterController).AddNode()
2020-07-26T17:21:31.7928883Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/master.go:262 +0x145
2020-07-26T17:21:31.7929372Z   github.com/ovn-org/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller.NewMaster.func1()
2020-07-26T17:21:31.7929909Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/master.go:62 +0x66
2020-07-26T17:21:31.7930386Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/informer.(*eventHandler).syncHandler()
2020-07-26T17:21:31.7930887Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer.go:275 +0x4f6
2020-07-26T17:21:31.7931388Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/informer.(*eventHandler).processNextWorkItem.func1()
2020-07-26T17:21:31.7931881Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer.go:220 +0x145
2020-07-26T17:21:31.7932382Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/informer.(*eventHandler).processNextWorkItem()
2020-07-26T17:21:31.7932879Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer.go:237 +0x6e
2020-07-26T17:21:31.7933340Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/informer.(*eventHandler).runWorker()
2020-07-26T17:21:31.7933836Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer.go:188 +0x38
2020-07-26T17:21:31.7934294Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/informer.(*eventHandler).runWorker-fm()
2020-07-26T17:21:31.7934789Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer.go:187 +0x41
2020-07-26T17:21:31.7935013Z   k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
2020-07-26T17:21:31.7935542Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x6f
2020-07-26T17:21:31.7935766Z   k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
2020-07-26T17:21:31.7936296Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb3
2020-07-26T17:21:31.7936511Z   k8s.io/apimachinery/pkg/util/wait.JitterUntil()
2020-07-26T17:21:31.7937048Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x10d
2020-07-26T17:21:31.7937262Z   k8s.io/apimachinery/pkg/util/wait.Until()
2020-07-26T17:21:31.7937787Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x5a
2020-07-26T17:21:31.7937883Z 
2020-07-26T17:21:31.7938053Z Goroutine 16 (running) created at:
2020-07-26T17:21:31.7938210Z   testing.(*T).Run()
2020-07-26T17:21:31.7938434Z       /opt/hostedtoolcache/go/1.13.4/x64/src/testing/testing.go:960 +0x651
2020-07-26T17:21:31.7938611Z   testing.runTests.func1()
2020-07-26T17:21:31.7938823Z       /opt/hostedtoolcache/go/1.13.4/x64/src/testing/testing.go:1202 +0xa6
2020-07-26T17:21:31.7938987Z   testing.tRunner()
2020-07-26T17:21:31.7939205Z       /opt/hostedtoolcache/go/1.13.4/x64/src/testing/testing.go:909 +0x199
2020-07-26T17:21:31.7939369Z   testing.runTests()
2020-07-26T17:21:31.7939591Z       /opt/hostedtoolcache/go/1.13.4/x64/src/testing/testing.go:1200 +0x521
2020-07-26T17:21:31.7971436Z   testing.(*M).Run()
2020-07-26T17:21:31.7972042Z       /opt/hostedtoolcache/go/1.13.4/x64/src/testing/testing.go:1117 +0x2ff
2020-07-26T17:21:31.7972212Z   main.main()
2020-07-26T17:21:31.7972361Z       _testmain.go:44 +0x223
2020-07-26T17:21:31.7972455Z 
2020-07-26T17:21:31.7972626Z Goroutine 93 (running) created at:
2020-07-26T17:21:31.7973291Z   github.com/ovn-org/ovn-kubernetes/go-controller/pkg/informer.(*eventHandler).Run()
2020-07-26T17:21:31.7973814Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer.go:142 +0x3f6
2020-07-26T17:21:31.7974439Z   github.com/ovn-org/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller.(*MasterController).Run.func1()
2020-07-26T17:21:31.7974961Z       /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/master.go:139 +0x6c
2020-07-26T17:21:31.7975134Z ==================
```

**Comment by dcbw** [comment]
@danwinship I think this is a more general testcase race issue, not specific to your PR. Retest perhaps?

**Comment by girishmg** [comment]
Argh. Stupid GoLand IDE  + GitHub integration. I did something on the IDE and it closed the PR itself. Sorry about that @danwinship @dcbw

**Comment by girishmg** [comment]
@danwinship this might not work for SharedGateway Mode. Let me re-look at your changes.

**Comment by danwinship** [comment]
The patch is supposed to have no effect on shared gateway mode. The whole point of the patch is that local gateway mode and shared gateway mode need to be handled differently, and we weren't doing that before.
